Implement projectExistentialAndUnwrapClass

augusto2112 opened the pull request:
* **Explanation**: Adds a new function, `projectExistentialAndUnwrapClass`, which behaves like `projectExistential` but is tailored for LLDB, so LLDB can find the dynamic type of objects using reflection.
* **Scope**: The new function behaves like `projectExistential `, except for the following differences:

    1 - When it comes to existentials, LLDB stores the address of the error
  pointer, which must be dereferenced before further processing.

    2 - When the existential wraps a class type, LLDB expects the address
  returned is the class instance itself and not the address of the
  reference.

* **Risk**: Low, this change is purely additive.
* **Testing**: Modified /validation-test/Reflection/existentials.swift to test the new functionality.
* **Issue**: rdar://55412978
* **Reviewer**: Adrian Prantl (@adrian-prantl)

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
     }
   }
 
+  bool
```
I understand that `projectExistential()` uses a similar API, but if we are adding a new API anyway, why not make this
```
llvm::Optional<std::pair<const TypeRef *, RemoteAddress>> getDynamicTypeAndAddressClassExistential(RemoteAddress object)
```
?

It might also be nice to add a Doxygen comment to function that explains what it does.

augusto2112 replied:
Sure! I prefer this style as well. 

About the Doxygen comment, I wrote one but it ended up being just repeating the function's name, so I think it might not add much, what do you think?

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+    return true;
+  }
+
+  bool getDynamicTypeAndAddressErrorExistential(
```
same comment as above :-)

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+  }
+
+  bool
+  getDynamicTypeAndAddressOpaqueExistential(RemoteAddress object,
```
and again

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+    }
+  }
+
+  bool projectExistentialForLLDB(RemoteAddress ExistentialAddress,
```
Here it would be really great to have a doxygen comment that explains the difference between this function and `projectExistential()`.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+    auto IsClass = [](const TypeRef *TypeResult) {
+      // When the existential wraps a class type, LLDB expects that the
+      // address returned is the class instance itself and not the address
+      // of the reference.
```
Could this extra unwrapping be done in LLDB, or is it nested deep inside this function? Basically I wonder if we could get by with a single API here or if that isn't feasible.

augusto2112 replied:
I don't see a clean way of doing this without altering the API of the original function, which is why I think keeping it as two separate functions might be the best option.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
+                                 const TypeRef *ExistentialTR,
+                                 const TypeRef **OutInstanceTR,
+                                 RemoteAddress *OutInstanceAddress) {
+    auto IsClass = [](const TypeRef *TypeResult) {
```
Since we expect TypeResult to be non-null, we might as well pass in a `const TypeRef &` here?

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
-      *OutInstanceTR = InstanceTR;
-      *OutInstanceAddress = OptMetaAndValue->PayloadAddress;
+      }
+      if (IsClass(*OutInstanceTR)) {
```
nit: in the LLVM style single-statement then-blocks don't get wrapped in { }.

adrian-prantl commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,61 +789,150 @@ class ReflectionContext
-    if (ExistentialTR == nullptr)
+    auto ExistentialRecordTI =
+        getRecordTypeInfo(ExistentialTR, ExternalTypeInfo);
+    if (ExistentialRecordTI == nullptr)
```
if (!ExistentialRecordTI)

adrian-prantl commented:
This looks good, it would be great if we could also add a test similar to the other reflection tests in the validation test suite.

augusto2112 commented:
@swift-ci test

adrian-prantl commented:
Please make sure to also cherry-pick this to the `main` branch.

augusto2112 commented:
Seems like macOS testing hung after running all tests?

```
15:17:03 PASS: Swift(macosx-x86_64) :: stdlib/StringNormalization.swift (14615 of 14618)
15:17:09 PASS: Swift(macosx-x86_64) :: stdlib/UnicodeTrieGenerator.gyb (14616 of 14618)
15:17:47 PASS: Swift(macosx-x86_64) :: stdlib/StringMemoryTest.swift (14617 of 14618)
16:41:07 
Build timed out (after 83 minutes). Marking the build as failed.
16:41:08 Build was aborted
16:41:08 Archiving artifacts
16:41:08 ninja: build stopped: interrupted by user.
16:41:12 Recording test results
16:41:14 ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?

```

Doesn't seems to be related to the test I changed:

```
15:04:54 PASS: Swift(macosx-x86_64) :: Reflection/existentials.swift (3675 of 14618)
```

augusto2112 commented:
@swift-ci test

adrian-prantl commented:
Looks like there's one real failure on 32-bit watchsimulator:

https://ci.swift.org/job/swift-PR-macos/27591/testReport/junit/Swift(watchsimulator-i386)/Reflection/existentials_swift/

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test

adrian-prantl commented:
We are going to need to swift-ci nominate this. Can you edit the PR description to include a block with

augusto2112 commented:
Since I refactored `projectExistential`, this change also affects remote mirrors as well (it returns the dynamic type for class existentials, which wasn't happening before). I _could_ leave `projectExistential` untouched here and open a second pr to swift/main with the refactor after this patch is merged.

adrian-prantl commented:
> Since I refactored `projectExistential`, this change also affects remote mirrors as well (it returns the dynamic type for class existentials, which wasn't happening before). I _could_ leave `projectExistential` untouched here and open a second pr to swift/main with the refactor after this patch is merged.

It is probably a good idea to move out the NFC refactoring into a separate patch anyway, so why not do that? It makes the decision to take the remaining patch easier when it has a smaller surface area.

augusto2112 committed:
Implement projectExistentialAndUnwrapClass

Implement a version of projectExistential tailored for LLDB. There are 2
differences when projecting existentials for LLDB:

1 - When it comes to existentials, LLDB stores the address of the error
    pointer, which must be dereferenced.
2 - When the existential wraps a class type, LLDB expects the address
    returned is the class instance itself and not the address of the
    reference.

This patch also adapts the swift reflection test machinery to test
projectExistentialAndUnwrapClass as well. This is done by exposing
the new functionality from swift reflection test.  It is tested in
existentials.swift, and ensures that the typeref information is
exactly the same as what is expected from projectExistential,
except the out address.

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci please nominate

augusto2112 commented:
@swift-ci test

Looks like I accidentally reversed one of the fixes I made to the `existentials.swift` test on 32 bits.

augusto2112 commented:
@swift-ci test

mikeash reviewed (approved):
LGTM. `DereferenceAndSet` confused me momentarily, but I think I like it now.

kastiglione commented on include/swift/SwiftRemoteMirror/SwiftRemoteMirror.h:
```diff
@@ -247,6 +247,19 @@ int swift_reflection_projectExistential(SwiftReflectionContextRef ContextRef,
+///   may derefence and set OutStartOfInstanceData if OutInstanceTypeRef is a 
+///   class TypeRef.
+SWIFT_REMOTE_MIRROR_LINKAGE
+int swift_reflection_projectExistentialAndUnwrapClass(
```
Where is this used?

augusto2112 replied:
This is used in `swift-reflection-test.c` so we can access it from the `existentials.swift` test.

kastiglione commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -789,6 +789,52 @@ class ReflectionContext
+    if (IsBridgedError)
+      *IsBridgedError = Result->IsBridgedError;
+
+    return {{std::move(TypeResult), Result->PayloadAddress}};
```
It might be helpful to declare `TypeResult` with its concrete type instead of `auto`? It's not immediately obvious why it's being moved.

augusto2112 replied:
You're right, this isn't necessary! I forgot to change this when editing the code.

augusto2112 commented:
@swift-ci test
